build.rs

```diff
@@ -33,17 +32,33 @@ fn main() {
                 None => "clang-16".into(),
             };
 
+            build.compiler(clang);
+        }
+
+        let compiler = build.get_compiler();
+        if compiler.is_like_clang() {
             build
-                .compiler(clang)
                 .no_default_flags(true)
                 .flag("--target=riscv64")
                 .flag("-march=rv64imc_zba_zbb_zbc_zbs");
-        } else {
+
+            if env::var("DEBUG").map(|v| v != "false").unwrap_or(false) {
+                build.flag("-g").flag("-fno-omit-frame-pointer");
+            }
+
+            let opt_level = env::var("OPT_LEVEL").expect("fetching OPT_LEVEL");
```

Comment:
There is a strange problem, I found that no matter what value the environment variable OPT_LEVEL is set to, it is always 0 in build.rs.

How to reproduce:

```rs
# Add this in build.rs
use std::io::Write;
let mut data_file = std::fs::File::create("/tmp/data.txt").unwrap();
data_file.write(opt_level.as_bytes()).unwrap();
```

And run

```sh
$ OPT_LEVEL=3 CLANG=clang-17 cargo build --features=build-with-clang --target=riscv64imac-unknown-none-elf --verbose
```

See also: https://github.com/rust-lang/cargo/issues/1226

Reply:
I don't think it's expected to set `OPT_LEVEL` by oneself, cargo might override this: `https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts`

My test is, when you use `--release` in `cargo build`, `OPT_LEVEL` will be 3, but when you skip `--release`, `OPT_LEVEL` will be 0.

It is a different topic if we want to enable customized `OPT_LEVEL`. It might just be better to use the following lines in `Cargo.toml`:

```
[profile.release]
opt-level = "s"
```

Reply:
Another improvement suggestion: if the OPT_LEVEL environment variable is not set, we should add a default -Os

Reply:
OK, the code here makes me think that the user can manually set the environment variable

Reply:
Yeah, it's about using the same optimization setting as in cargo profile.

Reply:
And FYI, the profile can be overridden per package, e.g.:

https://github.com/synapseweb3/forcerelay/blob/2f9ecfe5e4558639db9420554e25b5ee1ffd0315/Cargo.toml#L32-L34